core/src/oracles.ts

```diff
@@ -0,0 +1,135 @@
+     *   - split the received value into validity marker part and price part
+     * If the price is valid - return it, otherwise provide a NaN
+     **/
+    if (oracle.type !== 'CurrentAndNextPrice') {
+        return new BigNumber(NaN);
+    }
+    const slot = oracle.nextPriceSlotAddress;
+    const nextPriceFeed = await provider.getStorageAt(oracleAddress, slot);
+    const valueSplitPosition = oracle.slotPriceValueBeginsAtPosition;
+    const isPriceValid = nextPriceFeed.substring(0, valueSplitPosition);
+    if (parseInt(isPriceValid, 16) === 1) {
+        return new BigNumber(`0x${nextPriceFeed.substring(valueSplitPosition)}`);
+    }
+    return new BigNumber(NaN);
+};
+
+const currentPriceExtractors: Record<CollateralPriceSourceConfig['type'], CallableFunction> = {
+    CurrentPriceOnly: async (
+        oracle: OracleCurrentPriceOnly,
+        provider: ethers.providers.JsonRpcProvider,
+        oracleAddress: string
+    ) => {
+        /**
+         * Get current price from the contract via direct memory access (slot address)
+         * The price is stored in separate slot as a validity marker:
+         *   - the marker might be stored in the same slot as some other irrelevent value
+         *   - extract the byte that contains validity marker only
+         * If the price is valid - return it, otherwise provide a NaN
+         **/
+        const currentPriceFeed = await provider.getStorageAt(oracleAddress, oracle.currentPriceSlotAddress);
+        const valueSplitPosition = oracle.slotPriceValueBeginsAtPosition;
+        const storageValue = await provider.getStorageAt(oracleAddress, oracle.currentPriceValiditySlotAndOffset.slot);
+        const isPriceValid = parseInt(storageValue[oracle.currentPriceValiditySlotAndOffset.offset], 16) === 1;
+        return isPriceValid
+            ? new BigNumber(`0x${currentPriceFeed.substring(valueSplitPosition)}`)
+            : new BigNumber(NaN);
+    },
+    CurrentAndNextPrice: async (
+        oracle: OracleCurrentAndNextPrices,
+        provider: ethers.providers.JsonRpcProvider,
+        oracleAddress: string
+    ) => {
+        /**
+         * Determine if the current price can be fetched from the contract,
+         * If yes, fetch the price by direct memory access (via slot address)
+         * The price is stored in the same slot as a validity marker:
+         *   - split the received value into validity marker part and price part
+         * If the price is valid - return it, otherwise provide a NaN
+         **/
+        const currentPriceFeed = await provider.getStorageAt(oracleAddress, oracle.currentPriceSlotAddress);
+        const valueSplitPosition = oracle.slotPriceValueBeginsAtPosition;
+        const isPriceValid = parseInt(currentPriceFeed.substring(0, valueSplitPosition), 16) === 1;
+        return isPriceValid
+            ? new BigNumber(`0x${currentPriceFeed.substring(valueSplitPosition)}`)
+            : new BigNumber(NaN);
```

Comment:
This code seems to be same/similar to `getNextOraclePrice`. What's the difference? Can the code be reused?